haystack/nodes/preprocessor/merger.py

```diff
@@ -0,0 +1,448 @@
+                                   merged documents. Setting it to `False` always drops the page number from the
+                                   merged document.
+        :param tokens_count: The cumulative number of tokens contained in this group, if relevant.
+        """
+        if not group:
+            raise ValueError(
+                "No documents in the `group` parameter. "
+                "Make sure to pass some documents to the DocumentMerger.merge() method."
+            )
+
+        merged_content = separator.join([doc.content for doc in group])
+        merged_document = Document(content=merged_content)
+
+        # Realign headlines or erase them
+        headlines_meta = {}
+        if realign_headlines and any("headlines" in doc.meta.keys() for doc in group):
+            if any(doc.meta["headlines"] is not None for doc in group):
+                merged_headlines = merge_headlines(documents=group, separator=separator)
+                headlines_meta = {"headlines": merged_headlines}
+        else:
+            for doc in group:
+                if "headlines" in doc.meta.keys():
+                    del doc.meta["headlines"]
+
+        # Reset page number or erase it
+        page_number_meta = {}
+        if retain_page_number and any("page" in doc.meta.keys() for doc in group):
+            page_number_meta = {"page": min(int(doc.meta.get("page", inf)) for doc in group)}
+        else:
+            for doc in group:
+                if "page" in doc.meta.keys():
+                    del doc.meta["page"]
+
+        # Retain any other common key
+        merged_document.meta = {**common_values([doc.meta for doc in group]), **headlines_meta, **page_number_meta}
+
+        if tokens_count:
+            merged_document.meta["tokens_count"] = tokens_count
+
+        return merged_document
+
+
+def merge_headlines(documents: List[Document], separator: str) -> List[Dict[str, Union[str, int]]]:
+    """
+    Merges the headlines dictionary with the new position of each headline into the merged document.
+    Assumes the documents are in the same order as when they were merged.
+    """
+    aligned_headlines = []
+    position_in_merged_document = 0
+    for doc in documents:
+        for headline in deepcopy(doc.meta.get("headlines")) or []:
+            headline["start_idx"] += position_in_merged_document
+            aligned_headlines.append(headline)
+        position_in_merged_document += len(doc.content) + len(separator)
+    return aligned_headlines
+
+
+def common_values(list_of_dicts: List[Dict[str, Any]]) -> Dict[str, Any]:
+    """
+    Retains all keys shared across all the documents being merged.
+
+    Such keys are checked recursively, see tests.
+    """
+    merge_dictionary = deepcopy(list_of_dicts[0])
+    for key, value in list_of_dicts[0].items():
+
+        # if not all other dicts have this key, delete directly
+        if not all(key in dict.keys() for dict in list_of_dicts):
```

Comment:
To be on the safe side I think it would be good to rename the variable `dict` here to a variable that is not a built-in method of Python.